File: chess_java/src/chess/piecies/Piece.java
package chess.piecies;

import chess.Board;
import chess.Spot;
import chess.validators.PieceMoveValidator;
import chess.beans.ValidationResponse;

import java.util.List;

public abstract class Piece {

    private boolean killed = false;
    private boolean white = false;
    private PieceMoveValidator pieceMoveValidator;

    public Piece(boolean white)
    {
        this.setWhite(white);
    }

    public boolean isWhite()
    {
        return this.white == true;
    }

    public void setWhite(boolean white)
    {
        this.white = white;
    }

    public boolean isKilled()
    {
        return this.killed == true;
    }

    public void setKilled(boolean killed)
    {
        this.killed = killed;
    }

    public ValidationResponse canMove(Board board, Spot start, Spot end){
        return pieceMoveValidator.execute(board, start, end);
    }

    public abstract List<Spot> getPathSpots(Board board, Spot start, Spot end);

    public PieceMoveValidator getPieceMoveValidator() {
        return pieceMoveValidator;
    }

    public void setPieceMoveValidator(PieceMoveValidator pieceMoveValidator) {
        this.pieceMoveValidator = pieceMoveValidator;
    }

    public abstract void move(Board board, Spot start, Spot end);
}

File: chess_cpp/piecies/Piece.cpp
//
// Created by Hrach Khachatryan on 12/15/19.
//

#include "Piece.h"
#include "../beans/ValidationResponse.h"
#include "../validators/PieceMoveValidator.h"


Piece::Piece(bool white) {
    this->setWhite(white);
}

bool Piece::isWhite() {
    return this->white == true;
}

void Piece::setWhite(bool white) {
    this->white = white;
}

bool Piece::isKilled() {
    return this->killed == true;
}

void Piece::setKilled(bool killed) {
    this->killed = killed;
}

ValidationResponse *Piece::canMove(Board *board, Spot *start, Spot *end) {
    return pieceMoveValidator.execute(board, start, end);
}


PieceMoveValidator Piece::getPieceMoveValidator() {
    return pieceMoveValidator;
}

void Piece::setPieceMoveValidator(std::shared_ptr<PieceMoveValidator> pieceMoveValidator) {
    this->pieceMoveValidator = pieceMoveValidator;
}

std::vector<std::string> move(Board* board, Spot* start, Spot* end){
    std::vector<std::string> messages ;
    baseStateChangeChecker->execute(board, start, end, messages);

    end->setPiece(start->getPiece());
    start->setPiece(nullptr);

    return messages;
}



